contrib/smtp_proxy/filters/network/source/smtp_decoder.cc

```diff
@@ -0,0 +1,262 @@
+#include "contrib/smtp_proxy/filters/network/source/smtp_decoder.h"
+
+#include <vector>
+
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace SmtpProxy {
+
+static const absl::string_view CRLF = "\r\n";
+
+// While https://www.rfc-editor.org/rfc/rfc5321.html#section-4.5.3.1.4
+// prescribes a max command line of 512 bytes, individual SMTP
+// extensions may increase this. e.g. extensions that add an
+// additional keyword to the MAIL FROM: command typically increase the
+// max mail command by the length of that keyword. For BINARYMIME
+// (rfc3030) this is 16 bytes, for SMTPUTF8 (rfc6531), it is 8
+// bytes. To the letter of the rfc, you need to understand all of the
+// extensions advertised by the server to determine the exact
+// limit. This is not a full implementation of SMTP; the limit here is
+// meant as a circuit breaker against obvious garbage. The real
+// implementation on either side is free to enforce a more precise
+// limit.
+const size_t kMaxCommandLine = 1024;
+// Likewise, the rfc reply limit is 512 bytes
+// https://www.rfc-editor.org/rfc/rfc5321.html#section-4.5.3.1.5
+const size_t kMaxReplyLine = 1024;
+
+// The smtp proxy filter only expects to consume the first ~3
+// responses from the upstream (greeting, ehlo, starttls), which
+// should be much smaller than this but again, don't buffer an
+// unbounded amount.
+const size_t kMaxReplyBytes = 65536;
+
+Decoder::Result GetLine(Buffer::Instance& data, size_t start, std::unique_ptr<char[]>& raw_line,
+                        absl::string_view& line, size_t max_line) {
+  ssize_t crlf = data.search(CRLF.data(), CRLF.size(), start, max_line);
+  if (crlf == -1) {
+    if (data.length() >= 512) {
+      return Decoder::Result::Bad;
+    } else {
+      return Decoder::Result::NeedMoreData;
+    }
+  }
+  ASSERT(crlf >= 0);
+  // empty line is invalid
+  if (static_cast<size_t>(crlf) == start) {
+    return Decoder::Result::Bad;
+  }
+
+  size_t len = crlf + CRLF.size() - start;
+  ASSERT(len > CRLF.size());
+  raw_line.reset(new char[len]);
+  data.copyOut(start, len, raw_line.get());
+  line = absl::string_view(raw_line.get(), len);
+
+  return Decoder::Result::ReadyForNext;
+}
+
+Decoder::Result DecoderImpl::DecodeCommand(Buffer::Instance& data, Command& result) {
+  std::unique_ptr<char[]> raw_line;
+  absl::string_view line;
+  Result res = GetLine(data, 0, raw_line, line, kMaxCommandLine);
+  if (res != Result::ReadyForNext) {
+    return res;
+  }
+  result.wire_len = line.size();
+
+  const bool trailing_crlf = absl::ConsumeSuffix(&line, CRLF);
+  ASSERT(trailing_crlf);
```

Comment:
Is it possible that this assert fails if someone sends malformed payload?

Reply:
I intended it to be a postcondition of GetLine(): it only returns ReadyForNext if the line ends with crlf. I don't know what the general philosophy is in the Envoy code base for asserting on internal invariants like this. If we want to return an error here and shut down the connection, it needs a new "internal error" code instead of Decoder::Result::Bad which returns a perm error to the client.

Reply:
If you want to ASSERT on programming error (not runtime data error), then it is OK.